feat: add mutating FunctionDefinition functions

asterite opened the pull request:
# Description

## Problem

Part of #5668

## Summary

## Additional Context

I think we are missing a `FunctionDefinition::body`, but that could come later.

I had to get a `FuncMeta` many times in some functions because Rust wouldn't let me get two mutable borrows of interner.

Generics handling is missing... but I think right now generics can't appear in these functions, as attributes only run on top-level functions, not on impl or trait impl functions, but let me know if this is not the case.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

jfecher commented:
> Generics handling is missing... but I think right now generics can't appear in these functions, as attributes only run on top-level functions, not on impl or trait impl functions, but let me know if this is not the case.

Top-level functions can still have generics:
```rs
fn length_times_two<T, let N: u32>(array: [T; N]) -> u32 {
    array.len() * 2
}
```

jfecher commented on compiler/noirc_frontend/src/hir/comptime/interpreter/builtin.rs:
```diff
@@ -730,21 +756,206 @@ fn function_def_return_type(
+                }
+            })?;
+
+        let hir_pattern = interpreter.elaborate_item(interpreter.current_function, |elaborator| {
+            elaborator.elaborate_pattern_and_store_ids(
```
We don't want to resolve this parameter in the interpreter's current function - we want to resolve it in the target function.  So we should do `elaborate_item(Some(func_id), ...)`

This is probably the source of the "Non-comptime variable x referenced in comptime scope" error you mentioned over slack.

jfecher commented on compiler/noirc_frontend/src/hir/comptime/interpreter/builtin.rs:
```diff
@@ -730,21 +756,206 @@ fn function_def_return_type(
+            add_token_spans(parameter_name_tokens.clone(), parameters_argument_location.span);
+        let parameter_pattern =
+            parser::pattern().parse(parameter_name_quoted).map_err(|mut errors| {
+                let error = errors.swap_remove(0);
+                let rule = "a pattern";
+                InterpreterError::FailedToParseMacro {
+                    error,
+                    tokens: parameter_name_tokens,
+                    rule,
+                    file: location.file,
+                }
```
```suggestion
                InterpreterError::FailedToParseMacro {
                    error: errors.swap_remove(0),
                    tokens: parameter_name_tokens,
                    rule: "a pattern",
                    file: location.file,
                }
```

jfecher commented on compiler/noirc_frontend/src/hir/comptime/interpreter/builtin.rs:
```diff
@@ -730,21 +756,206 @@ fn function_def_return_type(
+            input_parameter,
+            parameters_argument_location,
+        )?;
+        let parameter_type = get_type(tuple.pop().unwrap(), parameters_argument_location)?;
+        let parameter_name_tokens = get_quoted(tuple.pop().unwrap(), parameters_argument_location)?;
+        let parameter_name_quoted =
+            add_token_spans(parameter_name_tokens.clone(), parameters_argument_location.span);
+        let parameter_pattern =
+            parser::pattern().parse(parameter_name_quoted).map_err(|mut errors| {
+                let error = errors.swap_remove(0);
+                let rule = "a pattern";
+                InterpreterError::FailedToParseMacro {
```
Looks like there's a lot of code in this loop. I wonder if we can clean it up any more?

E.g. I think:
1. Add token spans to quoted value
2. Parse token value with `[parser]`
3. Error with a `FailedToParseMacro`

Is a common pattern we can probably create a helper function for.

asterite replied:
Good idea, done!

jfecher commented on compiler/noirc_frontend/src/hir/comptime/interpreter/builtin.rs:
```diff
@@ -730,21 +756,206 @@ fn function_def_return_type(
+        Box::new(Type::Unit),
+    );
+
+    // TODO: generics. In Elaborator::define_function_meta there's this code:
+    //
+    //     if !generics.is_empty() {
+    //         typ = Type::Forall(generics, Box::new(typ));
+    //     }
```
Definitely need to handle generics! You should be able to take them from the existing function type. They may also be duplicated in `FuncMeta::all_generics` but I think those are in a different format.

asterite replied:
In the end I got the generics from `FuncMeta.typ`

jfecher commented on compiler/noirc_frontend/src/hir/comptime/interpreter/builtin.rs:
```diff
@@ -730,21 +756,206 @@ fn function_def_return_type(
+    ) = check_two_arguments(arguments, location)?;
+    let return_type = get_type(return_type_argument, return_type_argument_location)?;
+
+    let func_id = get_function_def(self_argument, self_argument_location)?;
+    let func_meta = interpreter.elaborator.interner.function_meta(&func_id);
+    match func_meta.function_body {
+        FunctionBody::Unresolved(_, _, _) => (),
+        FunctionBody::Resolving | FunctionBody::Resolved => {
+            return Err(InterpreterError::CannotMutateFunction { location })
+        }
+    }
```
This chunk of code is also repeated in each of the new functions, so we can have a helper for it.

jfecher commented on compiler/noirc_frontend/src/hir/comptime/interpreter/builtin.rs:
```diff
@@ -730,21 +756,206 @@ fn function_def_return_type(
+        }
+    }
+
+    let parameter_types = func_meta.parameters.iter().map(|(_, typ, _)| typ.clone()).collect();
+
+    let func_meta = interpreter.elaborator.interner.function_meta(&func_id);
+    let function_type =
+        Type::Function(parameter_types, Box::new(return_type.clone()), Box::new(Type::Unit));
+
+    // TODO: generics. In Elaborator::define_function_meta there's this code:
+    //
+    //     if !generics.is_empty() {
+    //         typ = Type::Forall(generics, Box::new(typ));
+    //     }
```
Probably easier to edit the existing function type

asterite replied:
I tried to edit the existing function but for some reason it didn't work, I'm not sure why.

asterite replied:
Oh, maybe it's because I also removed the `push_definition_type` call, but that should still be done. I'll try it again...

asterite replied:
It was that! It's working now.

asterite commented on compiler/noirc_frontend/src/hir/comptime/interpreter/foreign.rs:
```diff
@@ -31,10 +31,11 @@ fn poseidon2_permutation(
 
-    let (input, typ) = get_array(interner, input, location)?;
-    let state_length = get_u32(state_length, location)?;
+    let (input, typ) = get_array(interner, input, input_location)?;
+    let state_length = get_u32(state_length, state_length_location)?;
```
Or maybe `get_u32`, `get_array`, etc., could take a `(Value, Location)` tuple instead of two separate arguments. I'll try that approach first.

asterite replied:
Done. I think it looks okay, but let me know and I can change it back to `check_argument_count`. I mainly wanted to avoid having to pop arguments and get them in backwards order... but maybe it's not that confusing.

asterite commented:
Regarding generics, I wanted to add a test to prove that it works with generics, something like:

```rust
#[mutate_generic]
fn identity<T>() {}

comptime fn mutate_generic(f: FunctionDefinition) {
  // I want to change f to be `fn identity<T>(x: T) -> T { x }`
  f.set_parameters(&[quote { x }, ???]); // what type can I pass here?
```

jfecher commented:
Is this PR ready for review again? It is hard for me to tell because of the force push but I see all the comments are resolved

asterite commented:
Oh, yes, it is, sorry! Next time I'll ask for a re-review after pushing several commits.

I'll fix the conflicts now.

jfecher commented on compiler/noirc_frontend/src/hir/comptime/interpreter/builtin.rs:
```diff
@@ -915,3 +987,73 @@ pub(crate) fn extract_option_generic_type(typ: Type) -> Type {
     generics.pop().expect("Expected Option to have a T generic type")
 }
+
+fn check_function_not_yet_resolved(
```
Can you move these helpers to builtin_helpers.rs?

jfecher commented on compiler/noirc_frontend/src/hir/comptime/interpreter.rs:
```diff
@@ -44,6 +44,8 @@ mod builtin;
 mod foreign;
 mod unquote;
 
+pub(crate) type ValueAndLocation = (Value, Location);
+
```
```suggestion
```
I don't think the name `ValueAndLocation` adds anything over `(Value, Location)`

jfecher reviewed (commented):
Can you avoid force pushes after a review has been done in the future? It makes it a bit more difficult to review since I can't see changes since the last review any more and need to start over.

jfecher commented on compiler/noirc_frontend/src/hir/comptime/interpreter/builtin.rs:
```diff
@@ -915,3 +988,59 @@ pub(crate) fn extract_option_generic_type(typ: Type) -> Type {
     generics.pop().expect("Expected Option to have a T generic type")
 }
+
+fn parse<T>(
```
Can you move the other helper functions too? Here and below

asterite commented:
> Can you avoid force pushes after a review has been done in the future? It makes it a bit more difficult to review since I can't see changes since the last review any more and need to start over.

Sounds good!

I did a force push because we use squash-merge and this PR was on top of another PR. When the first PR merged, doing `git merge` gives conflicts. I found a simpler way to do that by using `git rebase --onto` to remove the extra commits. Is there another way to do this merge in a simpler way?

asterite commented:
(the conflicts are easy to solve, just always "keep mine", but it's a bit annoying to do it all the time for every stacked PR, so I wonder how you all deal with this)

jfecher commented:
@asterite usually when this happens I just go through the conflicts and "keep mine" or fix them in the case there are changes (e.g. changes requested from the previous PR).

jfecher reviewed (approved):
Thank you :+1:
